Sources/SwiftProtobuf/TextFormatScanner.swift

```diff
@@ -1109,54 +1106,239 @@ internal struct TextFormatScanner {
     ///
     /// This function accounts for as much as 2/3 of the total run
     /// time of the entire parse.
-    internal mutating func nextFieldNumber(names: _NameMap, messageType: any Message.Type) throws -> Int? {
+    internal mutating func nextFieldNumber(
+        names: _NameMap,
+        messageType: any Message.Type,
+        terminator: UInt8?
+    ) throws -> Int? {
+        while true {
+            skipWhitespace()
+            if p == end {
+                if terminator == nil {
+                    return nil
+                } else {
+                    // Never got the terminator.
+                    throw TextFormatDecodingError.malformedText
+                }
+            }
+            let c = p[0]
+            switch c {
+            case asciiLowerA...asciiLowerZ,
+                asciiUpperA...asciiUpperZ: // a...z, A...Z
+                let key = parseUTF8Identifier()
+                if let fieldNumber = names.number(forProtoName: key) {
+                    return fieldNumber
+                }
+                if !options.ignoreUnknownFields {
+                    throw TextFormatDecodingError.unknownField
+                }
+                // Unknown field name
+                break
+            case asciiOpenSquareBracket: // Start of an extension field
+                let key = try parseExtensionKey()
+                if let fieldNumber = extensions?.fieldNumberForProto(messageType: messageType, protoFieldName: key) {
+                    return fieldNumber
+                }
+                if !options.ignoreUnknownExtensionFields {
+                    throw TextFormatDecodingError.unknownField
+                }
+                // Unknown field name
+                break
+            case asciiOne...asciiNine:  // 1-9 (field numbers are 123, not 0123)
+                var fieldNum = Int(c) - Int(asciiZero)
+                p += 1
+                while p != end {
+                    let c = p[0]
+                    if c >= asciiZero && c <= asciiNine {
+                        fieldNum = fieldNum &* 10 &+ (Int(c) - Int(asciiZero))
+                    } else {
+                        break
+                    }
+                    p += 1
+                }
+                skipWhitespace()
```

Comment:
What is the limit on protobuf field numbers?   We should record `pstart = p` at the top of the above loop and then `p - pstart` is the total number of digits.  If there are too many digits, we should fail here with `unknownField`.

Reply:
fyi - Despite what the diff looks like, this isn't new code, it just got indented so it can loop over unknown field/values.

Fields are 1-536,870,911, so we could add logic to cap that.  I'm happy to do that in a follow up since it's not directly related to unknown fields.

Reply:
https://protobuf.dev/programming-guides/proto2/#assigning for the sepc

Reply:
I agree.  This does sound like a good follow-up task.